### DISCUSSION TITLE
Appear to be getting 2 instances of globals - one in the remix serverside code (action, loader) & one in the server.js hosting code. Does that make sense?

### DISCUSSION THREAD

**Discussion by Nick-Minutello** [remix-run/remix#9790] [Q&A] [answered]
Our express/koa server initialises/configures our logging (lg4js).
The logging instance is a global.

However, it appears that when we import the logger from our action code, we are getting an un-initialised instance. How is that possible?

Is it expected?

**Comment by gustavopch** [comment, marked as answer]
Yes, unfortunately, it's expected.

When you import the logger from your server file, you're importing it from the `app` folder. When you import it from other files, what's actually running is the code in the `build` folder.

So you actually have two loggers: the one in (for example) `app/logger.js` and the one in `build/server/index.js`. Both are being used when you run your app.

I believe you wouldn't have that problem if you used https://github.com/kiliman/remix-express-vite-plugin because the server file wouldn't exist.

I hope the Remix team is aware that this is confusing and can find a solution in the future.

**Reply by gustavopch** [reply]
@Nick-Minutello I found a simple solution to this — inspired by `remix-express-vite-plugin`. Basically, move as much as you can from `server.ts` to `entry.server.ts`.
```ts
// app/entry.server.ts
// ===================

// ... Existing code.

export const createApp = (build: ServerBuild, viteDevServer: ViteDevServer): Express => {
  const app = express()

  // ... Configure your Express app (including the middlewares that will use your logger).

  return app
}

// server.ts
// ===================

// ... Existing code.

const build = viteDevServer
  ? () => viteDevServer.ssrLoadModule('virtual:remix/server-build')
  : await import('./build/server/index.js' as string)

const app = (
  typeof build === 'function' ? await build() : build
).entry.module.createApp(build, viteDevServer)

app.listen(port, () => { ... })
```
The "parallel dimension" only happens in `server.ts`, so keep as little code in it as possible and move everything you can to `app/entry.server.ts` so your code is in the "same dimension" as all the rest of the code of your app.

**Comment by Nick-Minutello** [comment]
Ahh, actually, that makes sense. Many thanks.

We have elected to pass the log4js singleton in via the dependency injection (context).

**Reply by JWTappert** [reply]
Can you show an example of this? How are you using it later in your app, are you pulling it out of context?
